File: 018_4Sum.py

class question(object):

    def fourSum(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: List[List[int]]
        """
        if len(nums)< 4:
            return []

        nums.sort()

        s = []

        for i in range(len(nums)):
            if i > 0 and nums[i] == nums[i-1]:
            # skip the duplicate triples
                continue
            for j in range(i+1,len(nums)):
                if j>i+1 and nums[j]==nums[j-1]:
                    continue
                right = len(nums)-1
                left = j+1
                target_sub = target -nums[i]-nums[j]
                while left < right:
                    if nums[right]+nums[left]> target_sub:
                        right -=1
                    elif nums[right]+nums[left]< target_sub:
                        left += 1
                    else:
                        s.append([nums[i],nums[j],nums[left],nums[right]])
                        right -= 1
                        left += 1
                        while left< right and nums[right+1] == nums[right]:
                            right -= 1
                        while left< right and nums[left]== nums[left-1]:
                            left += 1
        return s



if __name__ == '__main__':

    inputs = [1, 0, -1, 0, -2, 2]

    k = question()
    print 'Input :', inputs
    r = k.fourSum(inputs,0)
    print  'Result : ',r

File: 018_4Sum.js
/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number[][]}
 */
var fourSum = function(nums, target) {
    if(nums.length<4){
        return([]);
    };
    var i= 0;
    var re_list = [];
    nums.sort(function(a,b){return(a-b)});
    console.log(nums);
    while(i<nums.length-3){
        if(i>0 && nums[i]===nums[i-1]){
            i++;
            continue;
        };
        var j = i+1;
        while(j<nums.length-2){
            if(j>i+1 && nums[j]===nums[j-1]){
                j++;
                continue;
            };
            var l=j+1, r=nums.length-1;
            while(l<r){
                var sum_cur = nums[i]+nums[j]+nums[l]+nums[r];
                if(sum_cur === target){
                    re_list.push([nums[i],nums[j],nums[l],nums[r]]);
                    l++;
                    if(l<r && nums[l]===nums[l-1]){
                        l++;
                    };
                    r--;
                    if(l<r &&nums[r]===nums[r+1]){
                        r--;
                    };
                }else if(sum_cur > target){
                    r--;
                }else{
                    l++;
                };
                console.log([i,j,l,r]);
            }
            j++;
        };
        i++;
    };
    return(re_list);
};

var string_example =[4,-9,-2,-2,-7,9,9,5,10,-10,4,5,2,-4,-2,4,-9,5];
var target = -13;
console.log(fourSum(string_example,target));
